src/RevCPU.cc

```diff
@@ -2651,95 +2654,102 @@ void RevCPU::UpdateThreadAssignments(uint32_t ProcID){
-                       Thread->GetThreadID(), ProcID);
-          break;
-        default: // Should DEFINITELY never happen
-          output.fatal(CALL_INFO, 99, "Error: Thread %" PRIu32 " on Core %" PRIu32 " is in an unknown state... This is a bug\n",
-                       Thread->GetThreadID(), ProcID);
-          break;
-        }
-        // State change has been handled, so clear the bit
-        Changes[HartID] = false;
+  // NOTE: At this point we handle EVERY thread that changed state
+  while( !Procs[ProcID]->GetThreadsThatChangedState().empty() ){
+    auto& Thread = Procs[ProcID]->GetThreadsThatChangedState().front();
+    // Handle the thread that changed state based on the new state
+    switch ( Thread->GetState() ) {
+    case ThreadState::DONE:
+      // This thread has completed execution
+      // We need to:
+      // 1. Remove it from the AssignedThreads map (The Hart will automatically be updated)
+      // 2. Move its ThreadID to the CompletedThreads list
+      output.verbose(CALL_INFO, 8, 0, "Thread %" PRIu32 " on Core %" PRIu32 " is DONE\n", Thread->GetThreadID(), ProcID);
+      AssignedThreads.at(ProcID).erase(Thread->GetThreadID());
+      CompletedThreads.emplace(Thread->GetThreadID());
+      if( AssignedThreads.at(ProcID).empty() ){
+        Enabled[ProcID] = false;
       }
-      else {
-        // This thread didn't experience a state change
-        continue;
+      break;
+    case ThreadState::BLOCKED:
+      // This thread is blocked (currently only caused by a rev_pthread_join)
+      // We need to:
+      // 1. Check if the thread it is waiting on has already completed
+      // 2. If it has... Thread can resume execution
+      // 3. If not, thread remains blocked
+      //    3a. Move its ThreadID to the BlockedThreads list
+      //    3b. Remove it from the AssignedThreads lis
+      output.verbose(CALL_INFO, 8, 0, "Thread %" PRIu32 "on Core %" PRIu32 " is BLOCKED\n", Thread->GetThreadID(), ProcID);
+      // -- 1.
+      if( ThreadCanProceed(Thread->GetThreadID()) ){
+        // -- 2.
+        output.verbose(CALL_INFO, 8, 0, "Thread %" PRIu32 " on Core %" PRIu32 " was waiting on thread %u which has already completed so it can proceed\n",
+                        Thread->GetThreadID(), ProcID, Thread->GetWaitingToJoinTID());
+        // Continue executing thread on same Core
+        Thread->SetState(ThreadState::RUNNING);
       }
+      else { // -- 3.
+        output.verbose(CALL_INFO, 8, 0, "Thread %" PRIu32 " on Core %" PRIu32 " was waiting on thread %u which has not yet completed so it remains blocked\n",
+                        Thread->GetThreadID(), ProcID, Thread->GetWaitingToJoinTID());
+        Thread->SetState(ThreadState::BLOCKED);
+        // -- 3a.
+        BlockedThreads.emplace(Thread->GetThreadID());
+
+        std::cout << Thread->to_string();
```

Comment:
Debug?

Reply:
as in debug state/mode? That might be handy for single stepping within the simulator

Reply:
Ah yeah, that shouldn't be there.

Reply:
> as in debug state/mode? That might be handy for single stepping within the simulator

"Debug?" meant "Is this a debug print statement left over which shouldn't be merged?"